compiler/symbol.go

```diff
@@ -373,21 +373,42 @@ func (c *compilerContext) checkWasmImport(f *ssa.Function, pragma string) {
 	}
 }
 
-// Check whether the type maps directly to a WebAssembly type, according to:
+// Check whether the type maps directly to a WebAssembly type.
+//
+// This reflects the relaxed type restrictions proposed here (except for structs.HostLayout):
+// https://github.com/golang/go/issues/66984
+//
+// This previously reflected the additional restrictions documented here:
 // https://github.com/golang/go/issues/59149
-func isValidWasmType(typ types.Type, isReturn bool) bool {
+func isValidWasmType(typ types.Type, isPointerOrField bool) bool {
 	switch typ := typ.Underlying().(type) {
 	case *types.Basic:
 		switch typ.Kind() {
-		case types.Int32, types.Uint32, types.Int64, types.Uint64:
+		case types.Bool:
+			return true
+		case types.Int, types.Uint, types.Int8, types.Uint8, types.Int16, types.Uint16, types.Int32, types.Uint32, types.Int64, types.Uint64:
 			return true
 		case types.Float32, types.Float64:
 			return true
-		case types.UnsafePointer:
-			if !isReturn {
-				return true
+		case types.Uintptr, types.UnsafePointer:
+			return true
+		case types.String:
+			return isPointerOrField
+		}
+	case *types.Array:
+		return isPointerOrField && isValidWasmType(typ.Elem(), true)
```

Comment:
This would allow arrays as struct fields, which doesn't seem to be allowed by the spec. I think this is an oversight in the spec because I don't see why arrays in struct fields would be disallowed while pointers to arrays are allowed.

I'd suggest ignoring this, or even better proposing a small change to the spec.

Reply:
I agree that this is an oversight in the spec, as long as the argument is a pointer to a struct, that struct should be fine to contain array elements. I will update the proposal.

Reply:
@johanbrandhorst thank you!